### PR TITLE
Add igaps reader

### PR DESCRIPTION
**Pull request by mcoughlin**
scripts to ingest igaps:
http://www.star.ucl.ac.uk/IGAPS/catalogue/

### CODE REVIEW

File: tools/ingest_igaps.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
+"""
+This tool will digest untarred VLASS source data from https://cirada.ca/vlasscatalogueql0 and ingest the data into Kowalski.
```
Comment (dmitryduev) [inline tools/ingest_igaps.py:2]: Needs updating.

File: tools/ingest_igaps.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+        "U2DeltaDEC",
+        "U2DetectionID",
+        "U2ErrBits",
+        "errBits2",
+    ]
+    hdulist = fits.open(file)
+    nhdu = 1
+    dataframe = pd.DataFrame(np.asarray(hdulist[nhdu].data), columns=names)
+
+    for chunk_index, dataframe_chunk in dataframe.groupby(
```
Comment (dmitryduev) [inline tools/ingest_igaps.py:230]: Why do you need to get data in chunks if you read in the whole thing into memory before this? To throttle db ingestion?

File: tools/ingest_igaps.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+        "U2Class",
+        "U2Seeing",
+        "U2MJD",
+        "U2DeltaRA",
+        "U2DeltaDEC",
+        "U2DetectionID",
+        "U2ErrBits",
+        "errBits2",
+    ]
+    hdulist = fits.open(file)
```
Comment (dmitryduev) [inline tools/ingest_igaps.py:226]: Maybe do some context management instead? Somewhat more idiomatic.

File: tools/ingest_igaps.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        log("Successfully disconnected from db")
+
+
+def run(
+    path: str = "./",
+    num_processes: int = 1,
+    batch_size: int = 2048,
+):
+    """Pre-process and ingest IGAPS catalog
+    :param path: path to CSV data files (~2.0 GB untarred)
```
Comment (dmitryduev) [inline tools/ingest_igaps.py:280]: Looks more like ~30 GB in total, but I didn't count exactly.

**Review by dmitryduev — APPROVED** [review]
LGTM! See, you're a certified K professional at this point :)

File: tests/test_tools.py [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 from ingest_ztf_source_features import run
```
Comment (dmitryduev) [inline tests/test_tools.py:3]: While you're at it, mind renaming this one into `run_ztf_source_features`?

File: tools/ingest_igaps.py (hunk 2 above)
Comment (mcoughlin) [reply tools/ingest_igaps.py:230]: Yes it seems to work nicely. I suppose I could remove it if desired.

File: tools/ingest_igaps.py (hunk 3 above)
Comment (mcoughlin) [reply tools/ingest_igaps.py:226]: Can you give an example?

File: tools/ingest_igaps.py [hunk 6]
```diff
@@ -1,5 +1,5 @@
 """
-This tool will digest untarred VLASS source data from https://cirada.ca/vlasscatalogueql0 and ingest the data into Kowalski.
+This tool will digest zipped IGAPS source data from http://www.star.ucl.ac.uk/IGAPS/catalogue/ and ingest the data into Kowalski.
```
Comment (dmitryduev) [inline tools/ingest_igaps.py:2]: So no need to unzip?

File: tools/ingest_igaps.py [hunk 7]
```diff
@@ -277,7 +277,7 @@ def run(
     batch_size: int = 2048,
 ):
     """Pre-process and ingest IGAPS catalog
-    :param path: path to CSV data files (~2.0 GB untarred)
+    :param path: path to CSV data files (~98 GB tarred)
```
Comment (dmitryduev) [inline tools/ingest_igaps.py:280]: really?

**Review by dmitryduev — APPROVED** [review]

File: tools/ingest_igaps.py (hunk 2 above)
Comment (dmitryduev) [reply tools/ingest_igaps.py:230]: No, leave it is as, was just curious.

File: tools/ingest_igaps.py (hunk 3 above)
Comment (dmitryduev) [reply tools/ingest_igaps.py:226]: I meant 
```python
with fits.open(file) as hdulist:
    ...
```

File: tools/ingest_igaps.py (hunk 6 above)
Comment (mcoughlin) [reply tools/ingest_igaps.py:2]: Nope!

File: tools/ingest_igaps.py (hunk 7 above)
Comment (mcoughlin) [reply tools/ingest_igaps.py:280]: http://www.star.ucl.ac.uk/IGAPS/catalogue/ I took 2.0 as characteristic for the individual files. The website claims 98 tottal.

File: tools/ingest_igaps.py [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+        "U2MJD",
+        "U2DeltaRA",
+        "U2DeltaDEC",
+        "U2DetectionID",
+        "U2ErrBits",
+        "errBits2",
+    ]
+    with fits.open(file) as hdulist:
+        nhdu = 1
+        dataframe = pd.DataFrame(np.asarray(hdulist[nhdu].data), columns=names)
```
Comment (dmitryduev) [inline tools/ingest_igaps.py:228]: I guess you need indentation here, you're not using hdulist below, so please revert the change below.

File: tools/ingest_igaps.py (hunk 6 above)
Comment (dmitryduev) [reply tools/ingest_igaps.py:2]: Nice!

File: tools/ingest_igaps.py (hunk 8 above)
Comment (mcoughlin) [reply tools/ingest_igaps.py:228]: you don't want
with fits.open(file) as hdulist
?

Comment (dmitryduev) [reply tools/ingest_igaps.py:228]: I do, I just think there's no need to place lines 229-onwards inside that block.

File: tools/ingest_igaps.py [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+                    log(str(e))
+                    bad_document_indexes.append(document_index)
+
+            if len(bad_document_indexes) > 0:
+                log("Removing bad docs")
+                for index in sorted(bad_document_indexes, reverse=True):
+                    del batch[index]
+
+            # ingest batch
+            mongo.insert_many(collection=collection, documents=batch)
```
Comment (dmitryduev) [inline tools/ingest_igaps.py:265]:
```suggestion
    for chunk_index, dataframe_chunk in dataframe.groupby(
        np.arange(len(dataframe)) // batch_size
    ):

        log(f"{file}: processing batch # {chunk_index + 1}")

        batch = dataframe_chunk.to_dict(orient="records")

        bad_document_indexes = []

        for document_index, document in enumerate(batch):
            try:
                # GeoJSON for 2D indexing
                document["coordinates"] = dict()
                # string format: H:M:S, D:M:S
                document["coordinates"]["radec_str"] = [
                    deg2hms(document["RA"]),
                    deg2dms(document["DEC"]),
                ]
                # for GeoJSON, must be lon:[-180, 180], lat:[-90, 90] (i.e. in deg)
                _radec_geojson = [document["RA"] - 180.0, document["DEC"]]
                document["coordinates"]["radec_geojson"] = {
                    "type": "Point",
                    "coordinates": _radec_geojson,
                }
            except Exception as e:
                log(str(e))
                bad_document_indexes.append(document_index)

        if len(bad_document_indexes) > 0:
            log("Removing bad docs")
            for index in sorted(bad_document_indexes, reverse=True):
                del batch[index]

        # ingest batch
        mongo.insert_many(collection=collection, documents=batch)
```

File: tools/ingest_igaps.py (hunk 8 above)
Comment (dmitryduev) [reply tools/ingest_igaps.py:228]: @kmshin1397 do you agree?

**Comment by dmitryduev** [comment]
@mcoughlin one more detail: could you please update the default config [here](https://github.com/dmitryduev/kowalski/blob/master/config.defaults.yaml#L324)? I.e. replace IPHAS with the appropriate stuff from IGAPS. Thanks!
Otherwise can go in as far as I'm concerned. Good job!

**Comment by mcoughlin** [comment]
@kmshin1397 can merge then once he is happy.

**Review by kmshin1397 — APPROVED** [review]
